.github/workflows/auto_pr_test.yml

```diff
@@ -0,0 +1,103 @@
+name: auto_pr_test
+
+# This action is triggered:
+# 1. when someone creates a pull request for a merge to the main branch
+# 2. when changes are merged into the main branch (via a pull request)
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+
+# Below are jobs, each of which runs sequentially.
+jobs:
+  # This job builds the box model and runs our test suite.
+  build:
+    runs-on: ${{ matrix.os }}
+
+    # A build matrix storing all desired configurations.
+    strategy:
+      matrix:
+        os: [ubuntu-latest] #, macos-latest]
+        build-type: [Debug] #, Release]
+
+    # Environment variables
+    env:
+      PETSC_ARCH: ${{ matrix.os }}
+      PETSC_GIT_HASH: a826417c4d
+      DEBUG: 1
+
+    # Steps for building and running tests.
+    steps:
+
+    - name: Installing compilers and tools (Linux)
+      if: contains(matrix.os, 'ubuntu')
+      run: sudo apt-get install cmake lcov gcc gfortran g++
+
+    - name: Installing compilers and tools (Mac)
+      if: contains(matrix.os, 'macos')
+      run: |
+        brew update
+        brew install git gcc lcov
+
+    - name: Checking out repository
+      uses: actions/checkout@v2
+
+    - name: Installing PETSc
+      run: |
+        git clone https://gitlab.com/petsc/petsc.git
+        cd petsc
+        git checkout $PETSC_GIT_HASH
+        export PETSC_DIR=$PWD
+        ./configure \
+        --with-cc=gcc \
+        --with-cxx=g++ \
+        --with-fc=gfortran \
+        --CFLAGS='-g -O0' --CXXFLAGS='-g -O0' --FFLAGS='-g -O0 -Wno-unused-function' \
+        --with-clanguage=c \
+        --with-debug=$DEBUG  \
+        --with-shared-libraries=0 \
+        --download-hdf5 \
+        --download-metis \
+        --download-parmetis \
+        --download-exodusii \
+        --download-netcdf \
+        --download-pnetcdf \
+        --download-zlib \
+        --download-fblaslapack \
+        --download-mpich=http://www.mpich.org/static/downloads/3.2/mpich-3.2.tar.gz
+        make all
```

Comment:
This will take a while. You could use an image that already has everything (libCEED uses `jedbrown/petsc:latest`, which is updated to track PETSc master periodically and there have been requests to automate it from PETSc upstream). Alternatively, use apt to install `libblis-dev` and `liblapack-dev`.

Also, why `--with-shared-libraries=0`? That increases link time because we have several executables.

In PETSc CI, we often build with debug, but with `COPTFLAGS=-O` (modest optimization; no debugging) because we aren't actually attaching a debugger to jobs in CI and thus no need to pay for debug symbols.

Reply:
Thanks for the image! That will help quite a bit.

I usually default to static linking because it's the only thing that Always Works (tm). In this case, though, we were doing this in Travis. Anyway, I agree about the link times and am happy to change it to dynamic.

I'll also add the `COPTFLAGS=0` argument.

Reply:
I built image this way so if you use it, you don't need to configure anything with PETSc.
https://github.com/jedbrown/jedbrown-dockerfiles/blob/main/petsc/Dockerfile

Do you happen to have experience with best practices for maintaining a collection of images, and in particular automating rebuilding/tagging/pushing several dependent images?

Reply:
The Docker-related stuff I've done is typically simple enough that I can accomplish rebuilding/tagging/pushing in relatively simple shell scripts. But I've not tried to express dependencies between images in any systematic way. Do you have anything specific in mind? Are you spending a lot of time waiting for certain images to rebuild?

Reply:
Sort of. I have one or more containers with mpich and dev tools, then PETSc, then adding other packages. `make` doesn't really track container results (because they aren't local files). I made a half-hearted attempt to use podman and buildah, but they seem to need more configuration to support non-root users and I don't know if it helps for my problem. Ideally there would be a tool that would look at `FROM` directives in Dockerfiles and admit a matrix of `ARG`s to apply.

Reply:
I'll poke around a bit and see if anyone has figured out anything that might help with this.